rockcraft/project.py

```diff
@@ -34,6 +37,105 @@ class ProjectValidationError(CraftError):
     """Error validatiing rockcraft.yaml."""
 
 
+class ArchitectureMapping(pydantic.BaseModel):
+    """Maps different denominations of the same architecture."""
+
+    description: str
+    deb_arch: str
+    compatible_uts_machine_archs: List[str]
+    go_arch: str
+
+
+_SUPPORTED_ARCHS: Dict[str, ArchitectureMapping] = {
+    "amd64": ArchitectureMapping(
+        description="Intel 64",
+        deb_arch="amd64",
+        compatible_uts_machine_archs=["amd64", "x86_64"],
+        go_arch="amd64",
+    ),
+    "arm": ArchitectureMapping(
+        description="ARM 32-bit",
+        deb_arch="armhf",
+        compatible_uts_machine_archs=["arm"],
+        go_arch="arm",
+    ),
+    "arm64": ArchitectureMapping(
+        description="ARM 64-bit",
+        deb_arch="arm64",
+        compatible_uts_machine_archs=["aarch64"],
+        go_arch="arm64",
+    ),
+    "i386": ArchitectureMapping(
+        description="Intel 386",
+        deb_arch="i386",
+        compatible_uts_machine_archs=["i386"],  # TODO: also include "i686", "x86_64"?
+        go_arch="386",
+    ),
+    "ppc64le": ArchitectureMapping(
+        description="PowerPC 64-bit",
+        deb_arch="ppc64el",
+        compatible_uts_machine_archs=["ppc64le"],
+        go_arch="ppc64le",
+    ),
+    "riscv64": ArchitectureMapping(
+        description="RISCV 64-bit",
+        deb_arch="riscv64",
+        compatible_uts_machine_archs=["riscv64"],
+        go_arch="riscv64",
+    ),
+    "s390x": ArchitectureMapping(
+        description="IBM Z 64-bit",
+        deb_arch="s390x",
+        compatible_uts_machine_archs=["s390x"],
+        go_arch="s390x",
+    ),
+}
+
+
+class Platform(pydantic.BaseModel):
+    """Rockcraft project platform definition."""
+
+    build_on: Optional[  # type: ignore
+        pydantic.conlist(str, unique_items=True, min_items=1)  # type: ignore
+    ] = pydantic.Field(alias="build-on")
+    build_for: Optional[  # type: ignore
+        pydantic.conlist(str, unique_items=True, min_items=1)  # type: ignore
+    ] = pydantic.Field(alias="build-for")
+
+    @pydantic.validator("build_for", pre=True)
+    @classmethod
+    def _vectorise_build_for(cls, val):
+        """Vectorise target architecture if needed."""
+        if isinstance(val, str):
+            val = [val]
+        return val
+
+    @pydantic.root_validator(skip_on_failure=True)
+    @classmethod
+    def _validate_platform_set(cls, values):
+        """Validate the build_on build_for combination."""
+        build_for = values["build_for"] if values.get("build_for") else []
```

Comment:
(minor)
```suggestion
        build_for = values.get("build_for", [])
```
Same for `build_on`

Reply:
unfortunately this won't work, because the YAML parser will serialize the values into a JSON like this: `{"build_for": None}`, so `values.get("build_for",[])` will in fact return `None` instead of []

Reply:
Yes you are right!